JDK9 will block setAccessible on modules

We're using `setAccessible(true)` to get reflective access to private API in a couple places. We cannot count on this after JDK9 is released.

In particular, `DefaultMethodHandler` in SQL Object is using `setAccessible` to get access to the `MethodHandle.Lookup` for private members, so that we can invoke the default ("super") method.

See:
- http://jigsaw-dev.1059479.n5.nabble.com/creating-proxies-for-interfaces-with-default-methods-td5711955.html
- http://mlvm-dev.openjdk.java.narkive.com/ocxrCtZV/proxy-an-interface-and-call-a-default-method

I've tested the latest JDBI snapshot on JDK 9-ea+136 and, unfortunately, default methods don't work.
```
java.lang.RuntimeException: java.lang.IllegalAccessException: access to public member failed: org.jdbi.v3.sqlobject.TestSqlObject$Dao.doesTransactionAnnotationWork()boolean/invokeSpecial, from org.jdbi.v3.sqlobject.TestSqlObject$Dao/2 (unnamed module @34b7bfc0)

	at org.jdbi.v3.sqlobject.DefaultMethodHandler.invoke(DefaultMethodHandler.java:63)
	at org.jdbi.v3.sqlobject.TransactionDecorator.lambda$invoke$0(TransactionDecorator.java:54)
	at org.jdbi.v3.core.transaction.LocalTransactionHandler.inTransaction(LocalTransactionHandler.java:173)
	at org.jdbi.v3.core.Handle.inTransaction(Handle.java:478)
	at org.jdbi.v3.sqlobject.TransactionDecorator.invoke(TransactionDecorator.java:57)
	at org.jdbi.v3.sqlobject.SqlObjectFactory.lambda$createInvocationHandler$18(SqlObjectFactory.java:241)
	at com.sun.proxy.$Proxy16.doesTransactionAnnotationWork(Unknown Source)
	at org.jdbi.v3.sqlobject.TestSqlObject.testTransactionAnnotationWorksOnInterfaceDefaultMethod(TestSqlObject.java:115)
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(java.base@9-ea/Native Method)
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(java.base@9-ea/NativeMethodAccessorImpl.java:62)
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(java.base@9-ea/DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(java.base@9-ea/Method.java:535)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.ExpectedException$ExpectedExceptionStatement.evaluate(ExpectedException.java:239)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:117)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:42)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:262)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:84)
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(java.base@9-ea/Native Method)
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(java.base@9-ea/NativeMethodAccessorImpl.java:62)
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(java.base@9-ea/DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(java.base@9-ea/Method.java:535)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:147)
Caused by: java.lang.IllegalAccessException: access to public member failed: org.jdbi.v3.sqlobject.TestSqlObject$Dao.doesTransactionAnnotationWork()boolean/invokeSpecial, from org.jdbi.v3.sqlobject.TestSqlObject$Dao/2 (unnamed module @34b7bfc0)
	at java.lang.invoke.MemberName.makeAccessException(java.base@9-ea/MemberName.java:915)
	at java.lang.invoke.MethodHandles$Lookup.checkAccess(java.base@9-ea/MethodHandles.java:1924)
	at java.lang.invoke.MethodHandles$Lookup.checkMethod(java.base@9-ea/MethodHandles.java:1864)
	at java.lang.invoke.MethodHandles$Lookup.getDirectMethodCommon(java.base@9-ea/MethodHandles.java:2013)
	at java.lang.invoke.MethodHandles$Lookup.getDirectMethodNoSecurityManager(java.base@9-ea/MethodHandles.java:2007)
	at java.lang.invoke.MethodHandles$Lookup.unreflectSpecial(java.base@9-ea/MethodHandles.java:1539)
	at org.jdbi.v3.sqlobject.DefaultMethodHandler.invoke(DefaultMethodHandler.java:51)
	... 38 more
```

--- a/sqlobject/src/main/java/org/jdbi/v3/sqlobject/DefaultMethodHandler.java
+++ b/sqlobject/src/main/java/org/jdbi/v3/sqlobject/DefaultMethodHandler.java
@@ -18,6 +18,7 @@ import static java.util.Collections.synchronizedMap;
 import java.lang.invoke.MethodHandle;
 import java.lang.invoke.MethodHandles;
 import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
 import java.lang.reflect.Method;
 import java.util.Map;
 import java.util.WeakHashMap;
@@ -25,22 +26,47 @@ import java.util.WeakHashMap;
 import org.jdbi.v3.core.extension.HandleSupplier;
 
 class DefaultMethodHandler implements Handler {
+    // MethodHandles.privateLookupIn(Class, Lookup) was added in JDK 9.
+    // JDK 9 allows us to unreflectSpecial() on an interface default method, where JDK 8 did not.
+    private static final Method PRIVATE_LOOKUP_IN = privateLookupIn();
+
+    private static Method privateLookupIn() {
+        try {
+            return MethodHandles.class.getMethod("privateLookupIn", Class.class, MethodHandles.Lookup.class);
+        } catch (NoSuchMethodException e) {
+            // Method was added in JDK 9
+            return null;
+        }
+    }
+
     private static final Map<Class<?>, MethodHandles.Lookup> privateLookups = synchronizedMap(new WeakHashMap<>());
 
     private static MethodHandles.Lookup lookupFor(Class<?> clazz) {
+        if (PRIVATE_LOOKUP_IN != null) {
+            try {
+                return (MethodHandles.Lookup) PRIVATE_LOOKUP_IN.invoke(null, clazz, MethodHandles.lookup());
+            } catch (IllegalAccessException | InvocationTargetException e) {
+                String message = String.format(
+                        "Error invoking MethodHandles.privateLookupIn(%s.class, MethodHandles.lookup()) in JDK 9+ runtime",
+                        clazz);
+                throw new RuntimeException(message, e);
+            }
+        }
+
+        // TERRIBLE, HORRIBLE, NO GOOD, VERY BAD HACK
+        // Courtesy of:
+        // https://rmannibucau.wordpress.com/2014/03/27/java-8-default-interface-methods-and-jdk-dynamic-proxies/
+
+        // We can use MethodHandles to look up and invoke the super method, but since this class is not an
+        // implementation of method.getDeclaringClass(), MethodHandles.Lookup will throw an exception since
+        // this class doesn't have access to the super method, according to Java's access rules. This horrible,
+        // awful workaround allows us to directly invoke MethodHandles.Lookup's private constructor, bypassing
+        // the usual access checks.
+
+        // This workaround is only used in JDK 8.x runtimes. JDK 9+ runtimes use MethodHandles.privateLookupIn()
+        // above.
         return privateLookups.computeIfAbsent(clazz, type -> {
             try {
-                // TERRIBLE, HORRIBLE, NO GOOD, VERY BAD HACK
-                // Courtesy of:
-                // https://rmannibucau.wordpress.com/2014/03/27/java-8-default-interface-methods-and-jdk-dynamic-proxies/
-
-                // We can use MethodHandles to look up and invoke the super method, but since this class is not an
-                // implementation of method.getDeclaringClass(), MethodHandles.Lookup will throw an exception since
-                // this class doesn't have access to the super method, according to Java's access rules. This horrible,
-                // awful workaround allows us to directly invoke MethodHandles.Lookup's private constructor, bypassing
-                // the usual access checks.
-
-                // We should get rid of this workaround as soon as a viable alternative exists.
 
                 final Constructor<MethodHandles.Lookup> constructor =
                         MethodHandles.Lookup.class.getDeclaredConstructor(Class.class, int.class);

